lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -414,19 +343,243 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
     return failure();
   }
 
-  LogicalResult transferWithinBlock(const LinearLayout &conversion,
-                                    ConvertLayoutOp op, OpAdaptor adaptor,
-                                    ConversionPatternRewriter &rewriter) const {
-    // TODO(jlebar): Implement me.
-    return failure();
+  LogicalResult
+  transferWithinBlockOrGroup(const LinearLayout &conversion, ConvertLayoutOp op,
+                             const LinearLayout &srcLayout,
+                             const LinearLayout &dstLayout, OpAdaptor adaptor,
+                             ConversionPatternRewriter &rewriter) const {
+    MLIRContext *ctx = op.getContext();
+    auto loc = op.getLoc();
+
+    assert(cvtNeedsSharedMemory(op.getSrc().getType(), op.getType()));
+
+    // TODO(jlebar): For now we handle only blocked/slice -> blocked/slice
+    // conversions.  Once we have ldmatrix support in
+    // load/storeDistributedToShared, we can remove this constraint.
+    std::function<bool(Attribute)> layoutIsOK = [&](Attribute layout) {
+      if (isa<BlockedEncodingAttr>(layout)) {
+        return true;
+      }
+      if (auto slice = dyn_cast<SliceEncodingAttr>(layout)) {
+        return layoutIsOK(slice.getParent());
+      }
+      return false;
+    };
+    if (!layoutIsOK(op.getSrc().getType().getEncoding()) ||
+        !layoutIsOK(op.getType().getEncoding())) {
+      return failure();
+    }
+
+    SmallVector<Value> inVals =
+        unpackLLElements(loc, adaptor.getSrc(), rewriter);
+    assert(!inVals.empty());
+
+    // Munge and unmunge are necessary because otherwise TargetInfo load/store
+    // cannot handle pointers correctly.
```

Comment:
I might say that the problem is that we canonically use MLIR's VectorTy, which cannot contain LLVM pointers.  (If we used LLVM's VectorTy it would work...)